### PR TITLE
Vulcan: Problem: TOC: Display like mobile causes list

### PR DESCRIPTION
**Pull request by ianrohde**
## Issue
The desktop-width TOC on problem pages is very convenient, but the lack of a "Causes" header and the icons shown on the mobile causes list give the page more of a "generic wiki feel". The problem page mockup has already been updated to reflect how we'd like the TOC to look: 

Image

[Mock-up](https://www.figma.com/file/QVLUEpDVoyiMM4MXEGq1L9/iFixit---Problems?type=design&node-id=296-44379&mode=design&t=AWzeP1tPVtFUcgqV-4)

_This one had a few moving parts. If I can help clarify anything, don't hesitate to reach out._

## CR/QA

`Troubleshooting/Refrigerator/Ice+Maker+Not+Making+Ice/481782` is a solid test-case with many causes and related problems.

<details>
<summary>Post-fix Video</summary>

https://github.com/iFixit/react-commerce/assets/1634505/90a123fc-d1f2-4df2-98f5-5070a77e7c6b

</details>

We want to make sure:

- Adds a "Causes" heading
- Adds headings to each TOC section (e.g. "Causes", etc)
- Adds icons in front of each section
  - Custom icon for "Introduction" and "Conclusions"
  - Numbers for solutions
  - Don't display "Related Problems" in the TOC _at all_ (not at any breakpoint, mobile or desktop)
    - We _do_ still want to show the Related Problems item in the causes list at the top of the page at mobile widths

Later additions:

- Add additional headings to TOC; "Troubleshooting" and "Conclusions"
- Make "Related Problems" sidebar sticky and scrollable to match TOC UX

Closes https://github.com/iFixit/ifixit/issues/50725

### CODE REVIEW

**Comment by kthaler** [comment]
Confirmed the following:

- [x] A "Causes" heading is now in the TOC
- [x] There are headings to each TOC section 
- [x] There are now icons in front of each section
- [x] "Introduction" and "Conclusions" have custom icons
- [x] Each solution displayed in the TOC side bar has the corresponding number displayed to the left of it
- [x] Related problems are no longer displayed in the TOC for mobile or desktop
- [x] Additional headings are added to TOC; "Troubleshooting" and "Conclusions"

QA 💎 - The new headers for the TOC look good.
deploy_block 🦖 - A few concerns

<details>
<summary>Concern 1</summary>
<br>

> We do still want to show the Related Problems item in the causes list at the top of the page at mobile widths

The Related Problems option is no longer visible in the TOC at mobile widths. EX:

Screenshot 2023-12-11 at 1 02 37 PM
</details>

<details>
<summary>Concern 2</summary>
<br>

> Make "Related Problems" sidebar sticky and scrollable to match TOC UX

I am getting a constant sidebar in the Related Problems section of `/Troubleshooting/Refrigerator/Ice+Maker+Not+Making+Ice/481782` that look a bit off in my opinion. The scroll bar is as long as the actual Related Problems section and feels unnecessary. 

https://react-commerce-git-vulcan-problem-toc-display-lik-62ed0b-ifixit.vercel.app/Troubleshooting/Refrigerator/Ice+Maker+Not+Making+Ice/481782

Screenshot 2023-12-11 at 12 51 27 PM

I am not seeing that issue here though:

https://react-commerce-git-vulcan-problem-toc-display-lik-62ed0b-ifixit.vercel.app/Troubleshooting/Refrigerator/Ice+Maker+Only+Crushing+Ice/481789

Screenshot 2023-12-11 at 12 51 41 PM
</details>

<details>
<summary>Concern 3</summary>
<br>

Finally, do we want the headings (Troubleshooting, Causes, Conclusions) in mobile to be the same as in desktop? For example, the everything in TOC on mobile in this screenshot is under the "Causes" heading:

Screenshot 2023-12-11 at 1 07 08 PM

while on desktop, there are 3 distinct headings:

Screenshot 2023-12-11 at 1 07 02 PM
</details>

**Comment by ianrohde** [comment]
Concern 2 and 3 are fixed. Concern 1 is going to take time

dev_block 🔴

**Comment by ianrohde** [comment]
@kthaler thanks for catching those 👍 

Concern 1 fixed in https://github.com/iFixit/react-commerce/pull/2143/commits/4830d9a70ec76093e3ee2efc27028fb6675debf6

un_dev_block 🟢

File: frontend/templates/troubleshooting/components/Causes.tsx [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+) {
+   const intros: (IntroData & TOCRecord)[] = useTOCContext<TOCEntry>()
+      .getItems()
+      .filter(isIntroData);
+   const solutions: (SolutionData & TOCRecord)[] = useTOCContext<TOCEntry>()
+      .getItems()
+      .filter(isCauseData);
+   const conclusions: (ConclusionData & TOCRecord)[] = useTOCContext<TOCEntry>()
+      .getItems()
+      .filter(isConclusionData);
```
Comment (jarstelfox) [inline frontend/templates/troubleshooting/components/Causes.tsx:42]: Minor: The `getItems` runs a sort so just calling it once ups the performance
```suggestion
   const items = useTOCContext<TOCEntry>().getItems();

   const intros: (IntroData & TOCRecord)[] = items.filter(isIntroData);
   const solutions: (SolutionData & TOCRecord)[] = items.filter(isCauseData);
   const conclusions: (ConclusionData & TOCRecord)[] = items.filter(isConclusionData);
```

File: frontend/templates/troubleshooting/components/Causes.tsx [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+function isCauseData(data: TOCEntry): data is SolutionData & TOCRecord {
+   return data.type === 'Solution';
+}
+
+function isConclusionData(data: TOCEntry): data is ConclusionData & TOCRecord {
+   return data.type === 'Conclusion';
+}
+
+export function Causes(
+   props: React.ComponentProps<typeof Box>,
```
Comment (jarstelfox) [inline frontend/templates/troubleshooting/components/Causes.tsx:31]: Chakra has a nice alias:

```suggestion
   props: BoxProps,
```

Comment (ianrohde) [reply frontend/templates/troubleshooting/components/Causes.tsx:31]: Looks like we need this as it was. Good looking out though

**Comment by deltuh-vee** [comment]
dev_block ✊ 
I'm seeing a lot of extra whitespace at the bottom of the [page](https://react-commerce-git-vulcan-problem-toc-display-lik-62ed0b-ifixit.vercel.app/Troubleshooting/Refrigerator/Ice+Maker+Not+Making+Ice/481782#Section_Software_Bugs) at roughly tablet size(820x1180).
CleanShot 2023-12-12 at 13 50 31

**Comment by ianrohde** [comment]
https://github.com/iFixit/react-commerce/pull/2143/commits/277e6aec3a80db7a79c5dbf7ef8a8b6abc5399c4 should resolve the issues you saw:

<details>

https://github.com/iFixit/react-commerce/assets/1634505/c520e19f-a12e-4a84-a0b7-3e81b079857b

</details>

un_dev_block 🟢

File: frontend/tests/jest/tests/__snapshots__/ProductListItem.test.tsx.snap [hunk 3]
```diff
@@ -8,10 +8,10 @@ exports[`ProductListItem renders and matches the snapshot 1`] = `
     role="group"
   >
     <div
-      class="css-mnl840"
+      class="css-11m6b36"
```
Comment (jarstelfox) [inline frontend/tests/jest/tests/__snapshots__/ProductListItem.test.tsx.snap:11]: I think this snapshot does not need to be updated?

**Review by jarstelfox — APPROVED** [review]
CR 🌵

File: frontend/tests/jest/tests/__snapshots__/ProductListItem.test.tsx.snap (hunk 3 above)
Comment (ianrohde) [reply frontend/tests/jest/tests/__snapshots__/ProductListItem.test.tsx.snap:11]: Branch was failing tests until I updated it. I thought it was odd too, but 🤷

**Comment by deltuh-vee** [comment]
QA 🎬
- Related Problems option is visible in the TOC at mobile widths.
- No longer getting unnecessary scrollbars for Related Problems.
- Troubleshooting and Conclusion Headers have been added to the sidebar TOC
- Repeated QA from https://github.com/iFixit/react-commerce/pull/2143#issuecomment-1850898463

**Comment by ianrohde** [comment]
deploy_block 🔴 in case Jarred has further input https://github.com/iFixit/react-commerce/pull/2143#discussion_r1424659373

**Comment by jarstelfox** [comment]
un_deploy_block 🌵 it's tests which should be safe enough to merge. I don;t suspect any issues. Let's just fix forward if required

**Comment by ianrohde** [comment]
Small bug on `/Troubleshooting/Garbage_Disposal/Hums/491105`. I'll open a follow-up to patch

**Pull request merged**